subprojects/wrapper/src/main/java/org/gradle/wrapper/Download.java

```diff
@@ -75,6 +77,8 @@ private void downloadInternal(URI address, File destination)
             addBasicAuthentication(address, conn);
             final String userAgentValue = calculateUserAgent();
             conn.setRequestProperty("User-Agent", userAgentValue);
+            conn.setConnectTimeout(CONNECTION_TIMEOUT_SECONDS);
+            conn.setReadTimeout(READ_TIMEOUT_SECONDS);
             in = conn.getInputStream();
```

Comment:
I think it would be nice to catch the `SocketTimeoutException` and produce a better error message by throwing a different exception with a more focused message, e.g. "Download timed out."